department-of-veterans-affairs/va-enp-api
# #95 -  year based partitioning notification table

# Description


issue #95 

#### **Feature: Year-Based Partitioning**
- Added range-based partitioning for the `Notification` table using PostgreSQL.
- Added event listener to created partitioned `Notification` tables after Notification created 
- Creates partitioned `Notification` tables from 2024 to next 10 years

#### **API Enhancements**
- Updated `/db/test` endpoint to include an optional `years` query parameter for filtering notifications.
- Returns both `Template` and `Notification` data, filtered by `years` if specified.

#### **Unit Tests**
- Added tests to verify partition creation with correct SQL and error handling.

**Note: Having queries use the proper partitioned table will be handled in a later user story. The SQLAlchemy does not gracefully handle querying on partition tables, although some functionality exist with simple queries using `WHERE created_at >= <date> AND created_at <= <date>` or using RAW SQL for partition tables.** 

A ticket is being created to gracefully handle ORM queries against the partitioned tables. 


## How Has This Been Tested?

1. Ran application locally in Docker and confirmed partitioned tables were created as expected. 
    <img width="366" alt="Screenshot 2025-01-06 at 2 38 38 PM" src="https://github.com/user-attachments/assets/a7935a22-73a9-40c5-abc4-98258d91feb3" />

2. Used DB testing endpoints `GET db/test?years=<years>` and POST `db/test` to test queries against the partitioned `Notification` table. 
    - Create Notification with different `created_at` timestamps thru endpoint. 10 years and 1 notification per year.  
    <img width="790" alt="Screenshot 2025-01-06 at 2 39 23 PM" src="https://github.com/user-attachments/assets/ed0c4aa3-2c93-49f1-9018-202e5f6d85f0" />

    - Filter for only `Notifications` with a `created_at` date of 2024
    <img width="914" alt="Screenshot 2025-01-06 at 2 40 22 PM" src="https://github.com/user-attachments/assets/0dc090bc-6258-4f2a-bd29-6547dfa732bb" />

4. Check Postgres SQL updates as expected. There should be one notification per year. 
    <img width="1370" alt="Screenshot 2025-01-06 at 2 41 39 PM" src="https://github.com/user-attachments/assets/6d316189-0ba0-4a2b-ac48-aed59a928052" />
    <img width="1336" alt="Screenshot 2025-01-06 at 2 42 36 PM" src="https://github.com/user-attachments/assets/01a15682-f957-46e7-95ee-73a4c359ca9b" />
    
5. Ran an EXPLAIN plan in Docker DB container 
```
EXPLAIN SELECT * FROM notifications WHERE created_at >= '2024-01-01' AND created_at < '2025-01-01';
```
```
 Bitmap Heap Scan on notifications_2024 notifications  (cost=4.19..12.66 rows=4 width=64)
   Recheck Cond: ((created_at >= '2024-01-01 00:00:00+00'::timestamp with time zone) AND (created_at < '2025-01-01 00:00:00+00'::timestamp with time zone))
   ->  Bitmap Index Scan on idx_notifications_2024_created_at  (cost=0.00..4.19 rows=4 width=0)
         Index Cond: ((created_at >= '2024-01-01 00:00:00+00'::timestamp with time zone) AND (created_at < '2025-01-01 00:00:00+00'::timestamp with time zone))
```

## Checklist

- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/notification-api/blob/main/.github/release.yaml) to the PR.
